// graphql-to-js-client-builder/index.js
/*
The MIT License (MIT)
Copyright (c) 2016 Shopify Inc.

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all
copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM,
DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR
OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE
OR OTHER DEALINGS IN THE SOFTWARE.


Version: 0.0.1 Commit: 76b290e
*/
(function (global, factory) {
	typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports, require('graphql/language'), require('babel-types'), require('babel-generator')) :
	typeof define === 'function' && define.amd ? define(['exports', 'graphql/language', 'babel-types', 'babel-generator'], factory) :
	(factory((global.GraphQLToJSClientBuilder = global.GraphQLToJSClientBuilder || {}),global.graphql_language,global.t,global.generate));
}(this, (function (exports,graphql_language,t,generate) { 'use strict';

generate = generate && 'default' in generate ? generate['default'] : generate;

// Recursive helper function for sortDefinitions
function visitFragment(fragment, fragments, fragmentsHash) {
  if (fragment.marked) {
    throw Error('Fragments cannot contain a cycle');
  }
  if (!fragment.visited) {
    fragment.marked = true;
    // Visit every spread in this fragment definition
    graphql_language.visit(fragment, {
      FragmentSpread(node) {
        // Visit the corresponding fragment definition
        visitFragment(fragmentsHash[node.name.value], fragments, fragmentsHash);
      }
    });
    fragment.visited = true;
    fragment.marked = false;
    fragments.push(fragment);
  }
}

// Sorts the definitions such that all fragment definitions are before operation
// definitions and fragments definitions are in reverse topological order
function sortDefinitions(definitions) {
  const fragments = definitions.filter(definition => {
    return definition.kind === 'FragmentDefinition';
  });

  // Set up a hash for quick lookup
  const fragmentsHash = {};

  fragments.forEach(fragment => {
    fragmentsHash[fragment.name.value] = fragment;
  });

  const operations = definitions.filter(definition => {
    return definition.kind === 'OperationDefinition';
  });

  const sortedFragments = [];

  fragments.forEach(fragment => {
    if (!fragment.visited) {
      visitFragment(fragment, sortedFragments, fragmentsHash);
    }
  });

  return sortedFragments.concat(operations);
}

function argValueToJS(argumentValue, clientVar) {
  switch (argumentValue.kind) {
    case 'StringValue':
      return t.stringLiteral(argumentValue.value);
    case 'EnumValue':
      return t.callExpression(t.memberExpression(clientVar, t.identifier('enum')), [t.stringLiteral(argumentValue.value)]);
    case 'IntValue':
      return t.numericLiteral(parseInt(argumentValue.value, 10));
    case 'FloatValue':
      return t.numericLiteral(parseFloat(argumentValue.value));
    case 'BooleanValue':
      return t.booleanLiteral(argumentValue.value);
    case 'ListValue':
      return t.arrayExpression(argumentValue.values.map(value => argValueToJS(value, clientVar)));
    case 'ObjectValue':
      return t.objectExpression(argumentValue.fields.map(field => {
        return t.objectProperty(t.identifier(field.name.value), argValueToJS(field.value, clientVar));
      }));
    case 'Variable':
      return t.callExpression(t.memberExpression(clientVar, t.identifier('variable')), [t.stringLiteral(argumentValue.name.value)]);
    default:
      throw Error(`Unrecognized argument value type "${argumentValue.kind}"`);
  }
}

function argsToJS(key, args, clientVar) {
  const argFields = args.map(argument => {
    return t.objectProperty(t.identifier(argument.name.value), argValueToJS(argument.value, clientVar));
  });

  return t.objectProperty(key, t.objectExpression(argFields));
}

function identifyOperation(selection, spreadsVar) {
  switch (selection.kind) {
    case 'Field':
      return {
        selectionConstructionArgs: [t.stringLiteral(selection.name.value)],
        operationMethodName: 'add',
        kind: 'Field'
      };
    case 'InlineFragment':
      return {
        selectionConstructionArgs: [t.stringLiteral(selection.typeCondition.name.value)],
        operationMethodName: 'addInlineFragmentOn',
        kind: 'InlineFragment'
      };
    default:
      return {
        selectionConstructionArgs: [t.memberExpression(spreadsVar, t.identifier(selection.name.value))],
        operationMethodName: 'addFragment',
        kind: 'Fragment'
      };
  }
}

function applyAlias(options, selection) {
  if (selection.alias) {
    options.push(t.objectProperty(t.identifier('alias'), t.stringLiteral(selection.alias.value)));
  }
}

function applyArguments(options, selection, clientVar) {
  if (!(selection.arguments && selection.arguments.length)) {
    return;
  }

  options.push(argsToJS(t.identifier('args'), selection.arguments, clientVar));
}

// Returns the body of the block statement representing the selections
function selectionSetToJS(selectionSet, parentSelectionName, spreadsVar, clientVar) {
  const selections = selectionSet.selections.map(selection => {
    var _identifyOperation = identifyOperation(selection, spreadsVar);

    const selectionConstructionArgs = _identifyOperation.selectionConstructionArgs,
          operationMethodName = _identifyOperation.operationMethodName,
          kind = _identifyOperation.kind;

    const fieldOptions = [];

    applyAlias(fieldOptions, selection);
    applyArguments(fieldOptions, selection, clientVar);

    // Add query options (i.e. alias and arguments) to the query
    if (fieldOptions.length) {
      selectionConstructionArgs.push(t.objectExpression(fieldOptions));
    }

    if (['Field', 'InlineFragment'].includes(kind) && selection.selectionSet) {
      const fieldNameOrTypeConstraint = selectionConstructionArgs[0].value;

      selectionConstructionArgs.push(selectionSetToJS(selection.selectionSet, fieldNameOrTypeConstraint, spreadsVar, clientVar));
    }

    return t.expressionStatement(t.callExpression(t.memberExpression(t.identifier(parentSelectionName), t.identifier(operationMethodName)), selectionConstructionArgs));
  });

  return t.arrowFunctionExpression([t.identifier(parentSelectionName)], t.blockStatement(selections));
}

function fragmentVisitor(jsNodes, clientVar, documentVar, spreadsVar) {
  return function visitor(node) {
    const selectionRootName = 'root';
    const fragmentDefinitionArguments = [t.stringLiteral(node.name.value), t.stringLiteral(node.typeCondition.name.value), selectionSetToJS(node.selectionSet, selectionRootName, spreadsVar, clientVar)];

    jsNodes.push(t.expressionStatement(t.assignmentExpression('=', t.memberExpression(spreadsVar, t.identifier(node.name.value)), t.callExpression(t.memberExpression(documentVar, t.identifier('defineFragment')), fragmentDefinitionArguments))));
  };
}

function typeConstraint(variableAst) {
  switch (variableAst.kind) {
    case 'NonNullType':
      return `${typeConstraint(variableAst.type)}!`;
    case 'ListType':
      return `[${typeConstraint(variableAst.type)}]`;
    default:
      return variableAst.name.value;
  }
}

function constructJSDefinition(variableAst, clientVar) {
  const variableConstructionArgs = [t.stringLiteral(variableAst.variable.name.value), t.stringLiteral(typeConstraint(variableAst.type))];

  if (variableAst.defaultValue) {
    variableConstructionArgs.push(argValueToJS(variableAst.defaultValue, clientVar));
  }

  return t.callExpression(t.memberExpression(clientVar, t.identifier('variable')), variableConstructionArgs);
}

function variableDeclarationsToJS(definitions, clientVar) {
  return t.arrayExpression(definitions.map(definition => constructJSDefinition(definition, clientVar)));
}

function applyName(graphQLNode, argList) {
  if (graphQLNode.name) {
    argList.push(t.stringLiteral(graphQLNode.name.value));
  }
}

function applyVariables(graphQLNode, argList, clientVar) {
  const definitions = graphQLNode.variableDefinitions;

  if (!(definitions && definitions.length)) {
    return;
  }

  argList.push(variableDeclarationsToJS(definitions, clientVar));
}

function operationFactoryFunction(graphQLNode) {
  switch (graphQLNode.operation) {
    case 'query':
      return 'addQuery';
    case 'mutation':
      return 'addMutation';
    default:
      throw new Error(`Operation: "${graphQLNode.operation}" is not currently supported`);
  }
}

function operationVisitor(jsNodes, clientVar, documentVar, spreadsVar) {
  return function visitor(node) {
    const selectionRootName = 'root';
    const operationDefinitionArgs = [];

    applyName(node, operationDefinitionArgs);
    applyVariables(node, operationDefinitionArgs, clientVar);

    operationDefinitionArgs.push(selectionSetToJS(node.selectionSet, selectionRootName, spreadsVar, clientVar));

    jsNodes.push(t.expressionStatement(t.callExpression(t.memberExpression(documentVar, t.identifier(operationFactoryFunction(node))), operationDefinitionArgs)));
  };
}

function extractFragmentDefinitons(definitions) {
  return definitions.filter(definition => definition.kind === 'FragmentDefinition');
}

function insertSpreadVar(nodes, identifier$$1) {
  nodes.push(t.variableDeclaration('const', [t.variableDeclarator(identifier$$1, t.objectExpression([]))]));
}

function declareDocument(nodes, clientVar, documentVar) {
  nodes.push(t.variableDeclaration('const', [t.variableDeclarator(documentVar, t.callExpression(t.memberExpression(clientVar, t.identifier('document')), []))]));
}

function documentToJSAst(graphQLAst, clientVar, documentVar, spreadsVar) {
  const jsGraphQLNodes = [];

  declareDocument(jsGraphQLNodes, clientVar, documentVar);

  const sortedgraphQLAst = Object.assign({}, graphQLAst, { definitions: sortDefinitions(graphQLAst.definitions) });
  const fragmentDefinitons = extractFragmentDefinitons(sortedgraphQLAst.definitions);

  if (fragmentDefinitons.length) {
    insertSpreadVar(jsGraphQLNodes, spreadsVar);
  }

  graphql_language.visit(sortedgraphQLAst, {
    FragmentDefinition: fragmentVisitor(jsGraphQLNodes, clientVar, documentVar, spreadsVar),
    OperationDefinition: operationVisitor(jsGraphQLNodes, clientVar, documentVar, spreadsVar)
  });

  return jsGraphQLNodes;
}

function transformToAst(graphqlCode, clientVar = 'client', documentVar = 'document', spreadsVar = 'spreads') {
  const graphQLAst = graphql_language.parse(graphqlCode);

  return documentToJSAst(graphQLAst, t.identifier(clientVar), t.identifier(documentVar), t.identifier(spreadsVar));
}

function transformToCode(graphqlCode, clientVar = 'client', documentVar = 'document', spreadsVar = 'spreads') {
  const jsAst = transformToAst(graphqlCode, clientVar, documentVar, spreadsVar);

  return `${generate(t.program(jsAst)).code}\n`;
}

exports.transformToAst = transformToAst;
exports['default'] = transformToCode;

Object.defineProperty(exports, '__esModule', { value: true });

})));
